os/kernel/binary_manager/binary_manager.c

```diff
@@ -0,0 +1,407 @@
+	bmvdbg("Load '%s' success! pid = %d\n", BIN_NAME(bin_idx), BIN_ID(bin_idx));
+
+	return OK;
+}
+
+/* Reboot the board */
+void binary_manager_board_reset(void)
+{
+#ifdef CONFIG_BOARD_ASSERT_AUTORESET
+	boardctl(BOARDIOC_RESET, EXIT_SUCCESS);
+#else
+	(void)irqsave();
+	for (;;) {
+#ifdef CONFIG_ARCH_LEDS
+		//board_autoled_on(LED_PANIC);
+		up_mdelay(250);
+		//board_autoled_off(LED_PANIC);
+		up_mdelay(250);
+#endif
+	}
+#endif
+}
+
+/* Create recovery thread to handle fault with pid */
+void binary_manager_recovery(int pid)
+{
+	int ret;
+	int argc;
+	char **argv;
+	char value[1];
+
+	bmvdbg("Try to recover fault with pid %d\n", pid);
+
+	if (pid > 0) {
+		argc = 1;
+		argv = (char **)kmm_malloc(sizeof(char *) * (argc + 1));
+
+		argv[0] = itoa(pid, (char *)&value, 10);
+		argv[1] = NULL;
+
+		ret = kernel_thread("recovery_thread", RECOVERYTHD_PRIORITY, RECOVERYTHD_STACKSIZE, recovery_thread, (char * const *)argv);
+		if (ret > 0) {
+			free(argv);
+			bmvdbg("Execute recovery thread with pid %d\n", pid);
+			return;
+		}
+	}
+
+	bmdbg("Failed to recover\n");
+
+	binary_manager_board_reset();
+}
+
+/* Get index in binary table with binary id */
+int binary_manager_get_index_with_binid(int binid)
+{
+	int bin_idx;
+
+	if (binid <= 0) {
+		bmdbg("Invalid binid %d\n", binid);
+		return ERROR;
+	}
+
+	for (bin_idx = 0; bin_idx < g_bin_count; bin_idx++) {
+		if (BIN_ID(bin_idx) == binid) {
+			bmdbg("FIND binary %d idx = %d\n", binid, bin_idx);
+			return bin_idx;
+		}
+	}
+
+	return ERROR;
+}
+
+/* Get binary info with binary name */
+int binary_manager_get_info_with_name(char *bin_name)
+{
+	return OK;
+}
+
+/* Get info of all registered binaries */
+int binary_manager_get_info_all(void)
+{
+	return OK;
+}
+/****************************************************************************
+ * Main Function
+ ****************************************************************************/
+int binary_manager(int argc, char *argv[])
+{
+	int ret;
+	int nbytes;
+	int bin_idx;
+	int load_cnt;
+	binmgr_request_t request_msg;
+	mqd_t binmgr_mq_fd;
+
+	struct mq_attr attr;
+	attr.mq_maxmsg = BINMGR_MAX_MSG;
+	attr.mq_msgsize = sizeof(binmgr_request_t);
+	attr.mq_flags = 0;
+
+	bmvdbg("Binary Manager STARTED\n");
+
+	/* Create binary manager message queue */
+	binmgr_mq_fd = mq_open(BINARYMGR_MSG_QUEUE, O_RDONLY | O_CREAT, 0666, &attr);
+	if (binmgr_mq_fd < 0) {		
+		bmdbg("Failed to open message queue\n");
+		return 0;
+	}
+
+	/* Load all user binaries */
+	load_cnt = 0;
+	for (bin_idx = 0; bin_idx < g_bin_count; bin_idx++) {
+		ret = binary_manager_load_binary(bin_idx);
+		if (ret == OK) {
+			load_cnt++;
+		}
+	}
+	bmvdbg("Load %d binaries!\n", load_cnt);
+
+	while (1) {
+		ret = ERROR;
+		bmvdbg("Wait for message\n");
+
+		nbytes = mq_receive(binmgr_mq_fd, (char *)&request_msg, sizeof(binmgr_request_t), NULL);
+		if (nbytes <= 0) {
+			bmdbg("receive ERROR %d, errno %d, retry!\n", nbytes, errno);
+			continue;
+		}
+
+		sched_lock();
```

Comment:
If you call sched_lock here, sometimes recovery thread cannot be made immediately.
Is it okay?